src/Frame/Frame.cc

```diff
@@ -1916,18 +1916,46 @@ void Frame::SaveFile(const std::string& root_folder, const CARTA::SaveFile& save
     fs::path output_filename(save_file_msg.output_file_name());
     fs::path directory(save_file_msg.output_file_directory());
     CARTA::FileType output_file_type(save_file_msg.output_file_type());
+    bool overwrite(save_file_msg.overwrite());
 
     // Set response message
     int file_id(save_file_msg.file_id());
     save_file_ack.set_file_id(file_id);
     bool success(false);
     casacore::String message;
 
+    if (output_filename.empty()) {
+        message = "Cannot save image with no filename.";
+        save_file_ack.set_success(success);
+        save_file_ack.set_message(message);
+        return;
+    }
+
     // Get the full resolved name of the output image
     fs::path temp_path = fs::path(root_folder) / directory;
     fs::path abs_path = fs::absolute(temp_path);
     output_filename = abs_path / output_filename;
 
+    if (fs::exists(output_filename)) {
+        if (output_filename.string() == in_file) {
+            message = "Cannot overwrite the source image.";
+        } else if (CasacoreImageType(output_filename.string()) == casacore::ImageOpener::UNKNOWN) {
+            // Not an image
+            if (fs::is_directory(output_filename)) {
+                message = "Cannot overwrite existing directory.";
+            }
+        } else if (!overwrite) {
+            // Image requires overwrite
+            message = "Cannot overwrite existing image.";
+        }
+
+        if (!message.empty()) {
+            save_file_ack.set_success(success);
+            save_file_ack.set_message(message);
+            return;
+        }
+    }
+
     if (output_filename.string() == in_file) {
```

Comment:
Is this part redundant now? Are there any circumstances under which this could ever be triggered if the file *doesn't* already exist? I guess someone could delete the file on disk while it is open in CARTA. But in that case we wouldn't be overwriting anything, right?

Reply:
(This is referring to the `if (output_filename.string() == in_file) {` block below the added code.)

Reply:
Good catch, this block was mistakenly left in.  I removed the redundant check.